R/read_csv.R

```diff
@@ -0,0 +1,239 @@
+#' if the run was run with save_warmup = 1.
+#'
+read_sample_csv <- function(output_files) {
+  sampling_info <- NULL
+  inverse_mass_matrix_diag <- c()
+  sampling_params_draws <- c()
+  post_warmup_draws_array <- c()
+  warmup_draws_array <- c()
+  for(output_file in output_files) {
+    # read meta data
+    if (is.null(sampling_info)) {
+      sampling_info <- read_sample_info_csv(output_file)
+      inverse_mass_matrix_diag <- sampling_info$inverse_mass_matrix_diag
+    } else {
+      csv_file_info <- read_sample_info_csv(output_file)
+      # check if sampling info matches
+      check_sampling_csv_info_matches(sampling_info,
+                                  csv_file_info)
+      sampling_info$id <- c(sampling_info$id,
+                            csv_file_info$id)
+      sampling_info$init <- c(sampling_info$init,
+                              csv_file_info$init)
+      sampling_info$random_seed <- c(sampling_info$random_seed,
+                                     csv_file_info$random_seed)
+      sampling_info$output_file <- c(sampling_info$output_file,
+                                     csv_file_info$output_file)
+      sampling_info$output_diagnostic_file <- c(sampling_info$output_diagnostic_file,
+                                                csv_file_info$output_diagnostic_file)
+      sampling_info$output_refresh <- c(sampling_info$output_refresh,
+                                        csv_file_info$output_refresh)
+      inverse_mass_matrix_diag <- cbind(inverse_mass_matrix_diag,
+                                        csv_file_info$inverse_mass_matrix_diag)
+    }
+    # read sampling data
+    if(sampling_info$sample_save_warmup == 1) {
+      num_of_draws <- sampling_info$sample_num_samples + sampling_info$sample_num_warmup
+    } else {
+      num_of_draws <- sampling_info$sample_num_samples
+    }
+    draws <- utils::read.csv(output_file, header = TRUE, comment.char = "#")
+    sampling_params_draws <- rbind(sampling_params_draws, draws[, sampling_info$sampler_params])
+    if(sampling_info$sample_save_warmup == 1) {
+      warmup_draws_array <- rbind(post_warmup_draws_array,
+                                  draws[1:sampling_info$sample_num_warmup, sampling_info$model_params])
+      post_warmup_draws_array <- rbind(post_warmup_draws_array,
+                                       draws[(sampling_info$sample_num_warmup+1):num_of_draws, sampling_info$model_params])
+    } else {
+      warmup_draws_array <- NULL
+      post_warmup_draws_array <- rbind(post_warmup_draws_array,
+                                       draws[, sampling_info$model_params])
+    }
+  }
+  #inverse mass matrix is returned separately
+  sampling_info$inverse_mass_matrix_diag <- NULL
+  if(!is.null(inverse_mass_matrix_diag)) {
+    dimnames(inverse_mass_matrix_diag) <- list(diagonal_elements = seq(dim(sampling_info$inverse_mass_matrix_diag)[1]),
+                                               chain_id = sampling_info$id)
```

Comment:
are dim names ok?

Reply:
Yeah but what happens if the inverse mass matrix isn't diagonal? 
Maybe it makes sense to have a list of length num_chains with each element the (diagonal) inv mass matrix for a chain? Does that make sense?

Reply:
Fixed, this now returns a list.
For "dense_e" it returns

```R
[[1]]
           [,1]       [,2]     [,3]      [,4]     [,5]       [,6]      [,7]      [,8]     [,9]     [,10]
 [1,] 10.625600 -0.1775010  5.13592  5.881030 10.24860  6.5330700  9.125610  9.287880  6.51366  8.112620
 [2,] -0.177501  0.7330090  2.78991  0.715517 -1.05897 -0.0550689 -1.234450 -0.103625  2.12444  1.546560
 [3,]  5.135920  2.7899100 49.15940  8.992190  2.56814  7.2243700  1.841220  8.513780 14.72390 12.334500
 [4,]  5.881030  0.7155170  8.99219 27.019900  5.13961  5.6690700  6.953020  5.209580  9.71617  8.217290
 [5,] 10.248600 -1.0589700  2.56814  5.139610 37.53400 10.9214000 11.363900  9.839640  5.07434  3.107000
 [6,]  6.533070 -0.0550689  7.22437  5.669070 10.92140 26.0761000  9.827830  6.162040  7.02962 -0.628198
 [7,]  9.125610 -1.2344500  1.84122  6.953020 11.36390  9.8278300 27.299000  6.840820  5.26091  0.301627
 [8,]  9.287880 -0.1036250  8.51378  5.209580  9.83964  6.1620400  6.840820 26.785600  4.63408  8.690570
 [9,]  6.513660  2.1244400 14.72390  9.716170  5.07434  7.0296200  5.260910  4.634080 31.13860  9.679820
[10,]  8.112620  1.5465600 12.33450  8.217290  3.10700 -0.6281980  0.301627  8.690570  9.67982 39.590800

[[2]]
           [,1]       [,2]     [,3]      [,4]       [,5]      [,6]      [,7]      [,8]     [,9]    [,10]
 [1,] 12.730900  0.5173660 13.04860 11.883800 10.9776000 10.686800  7.596150  9.280620 10.01750 10.30370
 [2,]  0.517366  0.6511760  2.45716  0.598146  0.0882897  0.876269 -0.218469  0.385524  2.52694  1.03454
 [3,] 13.048600  2.4571600 47.62290 12.420100 10.5577000 11.184100  8.432160  5.990020 17.20770 14.36120
 [4,] 11.883800  0.5981460 12.42010 30.238100 10.5635000  6.622850  6.984000 12.022500  8.43985 11.75660
 [5,] 10.977600  0.0882897 10.55770 10.563500 28.8693000  6.856210  8.481870 11.386100 10.79140 11.70670
 [6,] 10.686800  0.8762690 11.18410  6.622850  6.8562100 27.405100  6.880270  8.501930  8.44625  5.89534
 [7,]  7.596150 -0.2184690  8.43216  6.984000  8.4818700  6.880270 22.899300  6.683080  5.83925  8.69224
 [8,]  9.280620  0.3855240  5.99002 12.022500 11.3861000  8.501930  6.683080 33.118600 10.54110  7.44970
 [9,] 10.017500  2.5269400 17.20770  8.439850 10.7914000  8.446250  5.839250 10.541100 36.41820 13.79010
[10,] 10.303700  1.0345400 14.36120 11.756600 11.7067000  5.895340  8.692240  7.449700 13.79010 36.35480
```

and for diag_e it returns 

```R
[[1]]
 [1] 11.065100  0.835713 41.325200 22.856800 28.961600 28.468900 25.717100 24.657700 33.626300 29.015100

[[2]]
 [1] 11.950400  0.760574 42.447200 21.558300 24.298800 33.117300 23.900500 21.294900 29.486300 25.967800
```